### PR TITLE
SqlServerDsc: Added SQL Server 2019 Integration Tests (and unit test) plus additional, 'SqlSetup' resource `InstallSharedDir` fix.

### LINKED ISSUE #1310: SqlServerDsc: Add a new pipeline job to test SQL Server 2019 family
#### Details of the scenario you tried and the problem that is occurring
Suggest we add another job that can be run in the AppVeyor build worker to test the SQL Server 2019 family. For example SQL Server 2019, Reporting Services 2019, Power BI Reporting Services 2019. Or at least for the last two, maybe different builds than what are installed on the two existing jobs.

*SQL Server 2019 is not yet available, but as soon as they are we should look into this.*

#### Verbose logs showing the problem
n/a

#### Suggested solution to the issue
Add a new job in appveyor.yml and add the necessary changes to the SqlSetup, SqlRSSetup, and SqlPBIRSSetup (when available). 

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
n/a

#### SQL Server edition and version the target node is running
n/a

#### SQL Server PowerShell modules present on the target node
n/a

#### The operating system the target node is running
n/a

#### Version and build of PowerShell the target node is running
n/a

#### Version of the DSC module that was used ('dev' if using current dev branch)
Dev

### PR DESCRIPTION
**Pull request by SphenicPaul**
#### Pull Request (PR) description

Made a number of changes to enable SQL Server 2019 integration tests to be executed as part of the Azure DevOps, pipeline. 

Changes include:

- Amending the `SqlServerDsc` module, Azure DevOps pipeline (defined in `azure-pipelines.yml` definition) to add additional "Job" in the "Test" stage to run integration tests for SQL Server 2019. 
- Amended integration scripts for the `SqlAgentAlert`, `SqlAgentFailsafe`, `SqlAgentOperator`, `SqlDatabase`, `SqlDatabaseDefaultLocation`, `SqlDatabaseMail`, `SqlDatabaseObjectPermission`, `SqlDatabasePermission`, `SqlDatabaseUser`, `SqlEndpoint`, `SqlLogin`, `SqlProtocol`, `SqlProtocolTcpIp`, `SqlRS`, `SqlRSSetup`, `SqlReplication`, `SqlRole`, `SqlScript`, `SqlScriptQuery`, `SqlSecureConnection`, `SqlServiceAccount` and `SqlSetup` resources to support SQL Server 2019 execution.
- Added functionality (within integration tests) to download SQL Server 2019 ISO via an additional EXE.
- Updating integration test `README.md` file to include SQL Server 2019.
- Updated relevant unit tests, where required.
- Amended `SqlSetup` resource to output the correct `InstallSharedDir` value when used with SQL Server 2019 - This failure/omission was identified by the amended integration tests and resolved/fixed as part of this change.
- Updated a few other comments to reference SQL Server 2019

#### This Pull Request (PR) fixes the following issues

- Fixes #1310

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [x] Resource documentation updated in the resource's README.md.
- [ ] Resource parameter descriptions updated in schema.mof.
- [ ] Comment-based help updated, including parameter descriptions.
- [ ] Localization strings updated.
- [ ] Examples updated.
- [x] Unit tests updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Integration tests updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Code changes adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

---
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676)

### CODE REVIEW

File: tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1 [hunk 1]
```diff
@@ -37,6 +37,7 @@ try
         $previousProgressPreference = $ProgressPreference
         $ProgressPreference = 'SilentlyContinue'
 
+        # TODO: Does this need to be different between SQL2017 and SQL2019?
```
Comment (SphenicPaul) [inline tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1:40]: I'm unclear if this needs to be changed for SQL 2019? ... And if so, for/to what?

**Review by johlju — CHANGES_REQUESTED** [review]
Awesome work on these! Just small comments. :smiley: 

> Reviewed 33 of 35 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-:-MRnLl1Q53Lj3rLrJKIU:b-feirqs)* status: all files reviewed, 5 unresolved discussions (waiting on @SphenicPaul)

___
*[tests/Integration/DSC_SqlRS.config.ps1, line 26 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnI5DHBxcJZk0Cu5Ou:-MRnI5DI2fpt15zY76PV:bebwjoh) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRS.config.ps1#L26)):*
<details><summary><i alt="summary">Quoted 15 lines of code…</i></summary>

> ```
>     if($script:sqlVersion -eq '150')
>     {
>         # SQL2019
>         $instanceName = 'SSRS'
>         $isoImageName = 'SQL2019.iso'
>         # Additional variables required as ISO is downloaded via additional EXE
>         $downloadExeName = 'SQL2019_Download.exe'
>         $downloadIsoName = 'SQLServer2019-x64-ENU-Dev.iso'
>     }
>     elseif($script:sqlVersion -eq '140')
>     {
>         # SQL2017
>         $instanceName = 'SSRS'
>         $isoImageName = 'SQL2017.iso'
>     }
> ```
</details>


I think this is never used. Looking at the test now I think we should change the "dependencies-block" so this is not needed? See comment below.
___
*[tests/Integration/DSC_SqlRS.config.ps1, line 62 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnIcYD3XxG3xPBliQI:-MRnIcYD3XxG3xPBliQJ:bxhb5jh) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRS.config.ps1#L62)):*
> ```
>                 DownloadExePath      = "$env:TEMP\$downloadExeName"
>                 DownloadIsoPath      = "$env:TEMP\$downloadIsoName"
> ```

We can remove this, it looks like it is never used here.
___
*[tests/Integration/DSC_SqlRS.config.ps1, line 93 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnIlGgAG76NUM9a_8p:-MRnIlGgAG76NUM9a_8q:bp0a3o8) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRS.config.ps1#L93)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
>         MountImage 'MountIsoMedia'
>         {
>             ImagePath   = $Node.ImagePath
>             DriveLetter = $Node.DriveLetter
>             Ensure      = 'Present'
>         }
>         WaitForVolume 'WaitForMountOfIsoMedia'
>         {
>             DriveLetter      = $Node.DriveLetter
>             RetryIntervalSec = 5
>             RetryCount       = 10
>         }
> ```
</details>


This is only needed if `$script:sqlVersion -eq '130'`. We could move it into the if-block below right before `SqlSetup`, then we don't need the ISO's for the other versions in this test. What do you think?
___
*[tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1, line 4 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnK70TDp5tXtLTjm1L:-MRnK70TDp5tXtLTjm1M:brugvk3) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1#L4)):*
> ```
> 'Integration_SQL2019'
> ```

Maybe we should remove this as it downloads October 2017 executable? See next comment.
___
*[tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1, line 41 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnKWSxB2vVCyA8g1gs:-MRnKWSxB2vVCyA8g1gt:b3ww0t1) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1#L41)):*
> ```
>         $script:mockSourceMediaDisplayName = 'Microsoft SQL Server Reporting Services (October 2017)'
>         $script:mockSourceMediaUrl = 'https://download.microsoft.com/download/E/6/4/E6477A2A-9B58-40F7-8AD6-62BB8491EA78/SQLServerReportingServices.exe'
> ```

We could create and issue to add tests for newer (2019) reporting services. I know there are issues reported in the repos with newer RS when used with the resources, so might not be a quick fix adding an additional exe to download.

Also, maybe we should add separate pipeline jobs to test RS 2017 and newer? 

This can be done in another PR. Would you mind creating an issue for it so we can track it?

**Review by SphenicPaul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676)* status: 33 of 35 files reviewed, 5 unresolved discussions (waiting on @johlju)

___
*[tests/Integration/DSC_SqlRS.config.ps1, line 26 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnI5DHBxcJZk0Cu5Ou:-MRnhQsi0e-D52_t8wUE:b-896fix) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRS.config.ps1#L26)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
>     if($script:sqlVersion -eq '150')
>     {
>         # SQL2019
>         $instanceName = 'SSRS'
>         $isoImageName = 'SQL2019.iso'
>         # Additional variables required as ISO is downloaded via additional EXE
>         $downloadExeName = 'SQL2019_Download.exe'
>         $downloadIsoName = 'SQLServer2019-x64-ENU-Dev.iso'
>     }
>     elseif($script:sqlVersion -eq '140')
>     {
>         # SQL2017
>         $instanceName = 'SSRS'
>         $isoImageName = 'SQL2017.iso'
>     }
> ```

I think this is never used. Looking at the test now I think we should change the "dependencies-block" so this is not needed? See comment below.
</blockquote></details>

Done.
___
*[tests/Integration/DSC_SqlRS.config.ps1, line 62 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnIcYD3XxG3xPBliQI:-MRnhRoA5cxg007m1Dl7:b-896fix) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRS.config.ps1#L62)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
>                 DownloadExePath      = "$env:TEMP\$downloadExeName"
>                 DownloadIsoPath      = "$env:TEMP\$downloadIsoName"
> ```

We can remove this, it looks like it is never used here.
</blockquote></details>

Done.
___
*[tests/Integration/DSC_SqlRS.config.ps1, line 93 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnIlGgAG76NUM9a_8p:-MRnirGE4x5JbYo9r8pB:bcuas8u) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRS.config.ps1#L93)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
>         MountImage 'MountIsoMedia'
>         {
>             ImagePath   = $Node.ImagePath
>             DriveLetter = $Node.DriveLetter
>             Ensure      = 'Present'
>         }
>         WaitForVolume 'WaitForMountOfIsoMedia'
>         {
>             DriveLetter      = $Node.DriveLetter
>             RetryIntervalSec = 5
>             RetryCount       = 10
>         }
> ```

This is only needed if `$script:sqlVersion -eq '130'`. We could move it into the if-block below right before `SqlSetup`, then we don't need the ISO's for the other versions in this test. What do you think?
</blockquote></details>

Yes. That makes sense. 

I've now moved '[MountImage]MountIsoMedia' and '[WaitForVolume]WaitForMountOfIsoMedia' resource instances/definitions into the '130'-specific, if-block.
___
*[tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1, line 4 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnK70TDp5tXtLTjm1L:-MRnlMCr2VQIMOxFFxUN:batqmj8) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1#L4)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> 'Integration_SQL2019'
> ```

Maybe we should remove this as it downloads October 2017 executable? See next comment.
</blockquote></details>

Done.

I've removed 'SqlRSSetup' integration tests running as part of 'Integration_SQL2019' build and updated the comment to infer it has been done deliberately (rather than missed accidentally).

This was initially a question I'd raised then deleted as I was a little unclear if SSRS was directly tied into the SQL Server versions or not. Running the same tests for SQL2019 with the same EXE looked like it would provide little, additionally benefit but I'd updated more for consistency rather than anything.
___
*[tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1, line 41 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnKWSxB2vVCyA8g1gs:-MRnm0YHDHfZPq_qWAkw:bjoeyhe) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1#L41)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
>         $script:mockSourceMediaDisplayName = 'Microsoft SQL Server Reporting Services (October 2017)'
>         $script:mockSourceMediaUrl = 'https://download.microsoft.com/download/E/6/4/E6477A2A-9B58-40F7-8AD6-62BB8491EA78/SQLServerReportingServices.exe'
> ```

We could create and issue to add tests for newer (2019) reporting services. I know there are issues reported in the repos with newer RS when used with the resources, so might not be a quick fix adding an additional exe to download.

Also, maybe we should add separate pipeline jobs to test RS 2017 and newer? 

This can be done in another PR. Would you mind creating an issue for it so we can track it?
</blockquote></details>

I agree - I suspect if SSRS is more of a component in it's own right (and not directly tied into SQL versions), it makes more sense to test it in isolation from the rest of the SQL versions going forward so there's not a untidy, staggered coupling of these within the integration tests.

I'll raise an issue for this one shortly. 

I was also tempted to add a couple of other issues for SQL2012 and SQL2014 support in these integration tests given these versions could be used/around for a while...

[SQL2012 - Lifecycles](https://support.microsoft.com/en-us/lifecycle/search?alpha=sql%20server%202012) - 12th July 2022 (end of extended support of latest service pack)

[SQL2014 - Lifecycles](https://support.microsoft.com/en-us/lifecycle/search?alpha=sql%20server%202014) - 9th July 2024 (end of extended support of latest service pack)

Personally, I'd find SQL2014 tests useful (SQL2012 less so but could be useful to others), and unless there are problematic, functionality changes needing lots of refactoring/fixing, the changes aren't too significant in nature (if similar to adding SQL2019).

Let me know what you think.

**Review by SphenicPaul — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676)* status: 33 of 35 files reviewed, 5 unresolved discussions (waiting on @johlju)

___
*[tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1, line 41 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnKWSxB2vVCyA8g1gs:-MRnrByVDE4P0Dg8hb1s:b-k9qcl1) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1#L41)):*
> 12th July 2020
Correction: SQL2012 is 12th July 2022 (not 12th July 2020)

**Review by johlju — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-:-MRoTPYU-Q_bV5Ibjkrj:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

___
*[tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1, line 41 at r2](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-MRnKWSxB2vVCyA8g1gs:-MRoT0JaC6vM9DY8p9or:b382yzf) ([raw file](https://github.com/dsccommunity/sqlserverdsc/blob/d548265df050aaa72f4cf55ec215c88ed41eefcd/tests/Integration/DSC_SqlRSSetup.Integration.Tests.ps1#L41)):*
<details><summary><i>Previously, SphenicPaul (Paul J. Wilcox) wrote…</i></summary><blockquote>

> 12th July 2020
Correction: SQL2012 is 12th July 2022 (not 12th July 2020)
</blockquote></details>

Please add issues for SQL2012 and SQL2014 as well, agree they will be useful.

**Comment by johlju** [comment]
@SphenicPaul We need to skip the integration tests for SqlRS too for the 2019 job, it fails now since it does not install Reporting Services. Those tests will need to be added to the (separate) jobs for Reporting Services, and could be mentioned in the issue you are about to create.

**Review by SphenicPaul — COMMENTED** [review]
OK. I've removed the 'Integration_SQL2019' condition so the 'SqlRS', integration tests won't run in SQL2019, integration tests. I've also removed the '150' references to remove all SQL2019 'evidence'. 

Build running again now.

> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-:-MRo_wew6cGubVOLrXyC:b-fqaj6p)* status: 34 of 35 files reviewed, all discussions resolved (waiting on @johlju)

**Review by SphenicPaul — COMMENTED** [review]
Looks like the SQL2019 integration tests have now passed ... others still running.

> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-:-MRoiy3S9uAQztAy6BdF:b-imvit6)* status: 34 of 35 files reviewed, all discussions resolved (waiting on @johlju)

**Comment by SphenicPaul** [comment]
OK. I've now added the following issues off the back of this:

* #1677 - Add SQL 2014, integration tests
* #1678 - Add SQL 2012, integration tests
* #1679 - Add SSRS 2017 (onwards) integration tests

...feel free to add to these if you want to add more information/thoughts.

**Comment by johlju** [comment]
Thank you for adding the issues! I have nothing to add to them at the moment. 🙂 

Waiting for the tests to pass then I merge this.

**Review by johlju — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/sqlserverdsc/1676#-:-MRpHnFk9LA8j3cJU9jS:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**